### PR TITLE
feat(stream-model-instance-handler): Enable soft-delete

### PR DESCRIPTION
**Pull request by pawartur**
## Description

Enable soft-delete of MIDs by setting their content to `null` or `{}` and skipping attached model's schema validation in such case.

## How Has This Been Tested?

Describe the tests that you ran to verify your changes. Provide instructions for reproduction.

- [X] Added relevant tests

## PR checklist

Before submitting this PR, please make sure:

- [X] I have tagged the relevant reviewers and interested parties

### CODE REVIEW

File: packages/stream-model-instance-handler/src/model-instance-document-handler.ts [hunk 1]
```diff
@@ -204,6 +204,10 @@ export class ModelInstanceDocumentHandler implements StreamHandler<ModelInstance
    * @private
    */
   async _validateContent(context: Context, modelStreamId: StreamID, content: any): Promise<void> {
+    if (isEmpty(content)) {
+      // We leave the possibility to set content to null/{} as a way of 'soft' deletion of an MID
+      return Promise.resolve()
```
Comment (ukstv) [inline packages/stream-model-instance-handler/src/model-instance-document-handler.ts:209]:
```suggestion
      return
```

Comment (ukstv) [reply packages/stream-model-instance-handler/src/model-instance-document-handler.ts:209]: @pawartur does it mean that you could do "soft-delete", and then add content? Is it intended?

**Review by ukstv — CHANGES_REQUESTED** [review]

File: packages/stream-model-instance-handler/src/model-instance-document-handler.ts (hunk 1 above)
Comment (pawartur) [reply packages/stream-model-instance-handler/src/model-instance-document-handler.ts:209]: Good question! I assumed that this is what we meant when we were discussing it.

Is there a way that we can soft delete and prevent any following content changes from that point?

**Review by willex — APPROVED** [review]

File: packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts [hunk 2]
```diff
@@ -298,6 +298,18 @@ describe('ModelInstanceDocumentHandler', () => {
     await checkSignedCommitMatchesExpectations(did, commit, expectedGenesis)
   })
 
+  it('null content in genesis is valid', async () => {
```
Comment (stbrody) [inline packages/stream-model-instance-handler/src/__tests__/model-instance-document-handler.test.ts:301]: what's the use case for allowing null content in genesis?

File: packages/stream-model-instance-handler/src/schema-utils.ts [hunk 3]
```diff
@@ -1,6 +1,12 @@
 import ajv, { SchemaObject } from 'ajv/dist/2020.js'
 import addFormats from 'ajv-formats'
 
+export function isEmpty(obj: any): boolean {
```
Comment (stbrody) [inline packages/stream-model-instance-handler/src/schema-utils.ts:4]: this would return `false` if given `null` - is that the desired behavior?

Comment (stbrody) [reply packages/stream-model-instance-handler/src/schema-utils.ts:4]: I actually feel like we **shouldn't** allow setting to an empty document `{}` but only allow setting to an explicit null.  That is a more clear signal that the goal is to delete the stream.  Allowing an empty document could let through an application programming bug where the content object is built programmatically but winds up with no fields set.  Making the app explicitly set null means its less likely to happen by accident.

**Review by stbrody — CHANGES_REQUESTED** [review]
If the contents of a MID get set to null, we should probably take that as a sign to remove the document from the index as well so it doesn't show up in queries.

Also, rather than allowing setting content to `null` directly using the `replace()` method, I wonder if we should have a separate function in the MID API for doing these soft-deletes.  Would make things more clear and explicit.  The hard part of course is figuring out what we would name it.  `advisoryDelete()`?  `unIndex()`?

CC @oed @PaulLeCam

File: packages/stream-model-instance-handler/src/model-instance-document-handler.ts (hunk 1 above)
Comment (stbrody) [reply packages/stream-model-instance-handler/src/model-instance-document-handler.ts]: We certainly could change the commit handler to not allow setting contents if the current contents are null, but I'm not sure that we actually want to do that?

**Comment by pawartur** [comment]
Closing this PR. We've decided with @stbrody and @PaulLeCam that this needs to be properly designed and discussed before implementation, specifically with regard to how removing a soft-deleted MID from index dbs should work. We'll come back to this when we have time to properly design it.

**Pull request closed without merging**